es6-shim.js

```diff
@@ -2097,6 +2170,257 @@
+
+        if ('value' in desc) {
+          if (!desc.writable) {
+            return false;
+          }
+
+          if (!ES.TypeIsObject(receiver)) {
+            return false;
+          }
+
+          var existingDesc = Object.getOwnPropertyDescriptor(receiver, key);
+
+          if (existingDesc) {
+            return __defineOwnProperty(receiver, key, {
+              value: value
+            });
+          } else {
+            return __defineOwnProperty(receiver, key, {
+              value: value,
+              writable: true,
+              enumerable: true,
+              configurable: true
+            });
+          }
+        }
+
+        if (desc.set) {
+          desc.set.call(receiver, value);
+          return true;
+        }
+
+        return false;
+      };
+
+      // Some Reflect methods are basically the same as
+      // those on the Object global, except that a TypeError is thrown if
+      // target isn't an object. As well as returning a boolean indicating
+      // the success of the operation.
+      defineProperties(Reflect, {
+
+        // Apply method in a functional form.
+        apply: ES.Call,
+
+        // New operator in a functional form.
+        construct: function construct(constructor, args) {
+          if (!ES.IsCallable(constructor)) {
+            throw new TypeError('First argument must be callable.');
+          }
+
+          return ES.Construct(constructor, args);
+        },
+
+        defineProperty: wrapObjectFunction(Object.defineProperty),
+
+        // When deleting a non-existant or configurable property,
+        // true is returned.
+        // When attempting to delete a non-configurable property,
+        // it will return false.
+        deleteProperty: function deleteProperty(target, key) {
+          if (!ES.TypeIsObject(target)) {
+            throw new TypeError('target must be an object');
+          }
+
+          var desc = Object.getOwnPropertyDescriptor(target, key);
+
+          if (desc && !desc.configurable) {
+            return false;
+          }
+
+          // Will return true.
+          return delete target[key];
+        },
+
+        enumerate: function enumerate(target) {
+          if (!ES.TypeIsObject(target)) {
+            throw new TypeError('target must be an object');
+          }
+
+          return new ObjectIterator(target, 'key');
+        },
+
+        // Syntax in a functional form.
+        get: function get(target, key) {
+          if (!ES.TypeIsObject(target)) {
+            throw new TypeError('target must be an object');
+          }
+
+          var receiver = arguments.length > 2 ? arguments[2] : target;
+
+          return internal_get(target, key, receiver);
+        },
+
+        getOwnPropertyDescriptor: throwUnlessTargetIsObject(Object.getOwnPropertyDescriptor),
+
+        getPrototypeOf: throwUnlessTargetIsObject(Object.getPrototypeOf),
+
+        has: function has(target, key) {
+          if (!ES.TypeIsObject(target)) {
```

Comment:
Could we wrap `function has` in `throwUnlessTargetIsObject` instead, so it's consistent?